gcr.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	fluxapi "github.com/fluxcd/flux/pkg/api"
+)
+
+const GoogleContainerRegistry = "GoogleContainerRegistry"
+const insert = "INSERT"
+const tokenIndex = len("Bearer ")
+
+type data struct {
+	Action string `json:"action"`
+	Digest string `json:"digest"`
+	Tag    string `json:"tag"`
+}
+
+type payload struct {
+	Message struct {
+		Data         string    `json:"data"`
+		MessageID    string    `json:"messageId"`
+		PublishTime  time.Time `json:"publishTime"`
+		Subscription string    `json:"subscription"`
+	} `json:"message"`
+}
+
+type auth struct {
+	Aud string `json:"aud"`
+}
+
+func init() {
+	Sources[GoogleContainerRegistry] = handleGoogleContainerRegistry
+}
+
+func handleGoogleContainerRegistry(s fluxapi.Server, _ []byte, w http.ResponseWriter, r *http.Request, config Endpoint) {
+	// authenticate based on config
+	if config.Authentication.Enabled {
+		if err := authenticateRequest(&http.Client{}, r.Header.Get("Authorization"), config.Authentication.Audience); err != nil {
+			http.Error(w, "Cannot authorize request", http.StatusOK)
+			log(GoogleContainerRegistry, err.Error())
+			return
+		}
+	}
+
+	var p payload
+	if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
+		http.Error(w, "Cannot decode payload", http.StatusOK)
+		log(GoogleContainerRegistry, err.Error())
+		return
+	}
+
+	raw, _ := base64.StdEncoding.DecodeString(p.Message.Data)
+
+	var d data
+	json.Unmarshal(raw, &d)
+
+	if strings.ToUpper(d.Action) != insert {
```

Comment:
I did a double-take at `.ToUpper` vs. `insert` here. Perhaps either make the const value "insert" and compare it to `.ToLower`, or make the const name `INSERT` (though go lint might take offence at that, not sure).